File: 프로그래머스/2/12952. N－Queen/N－Queen.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution {
    int ans;
    int N;
    int[] col;
    int[] r_diag;
    int[] l_diag;

    public void change(int r, int c, int v) {
        col[c] = v;
        r_diag[r+c] = v;
        l_diag[r-c+N] = v;      
    }
    
    public void backtrack(int r_idx) {
        if (r_idx == N) {
            ans++;
            return;
        }
        for (int c_idx=0; c_idx<N; ++c_idx) {
            if (col[c_idx] == 0 && r_diag[r_idx+c_idx] == 0 && l_diag[r_idx-c_idx+N] == 0) {
                change(r_idx, c_idx, 1);
                backtrack(r_idx+1);
                change(r_idx, c_idx, 0);
            }
        }
    }    
    
    public int solution(int n) {
        ans = 0;
        N = n;
        col = new int[n];
        r_diag = new int[2*n];
        l_diag = new int[2*n];
        
        backtrack(0);        
        
        return ans;
    }
}

File: 프로그래머스/2/12952. N－Queen/N－Queen.js
function solution(n) {
    let answer = 0;
    
    const R = Array.from({length: n}, ()=>false);
    const C = Array.from({length: n}, ()=>false);
    const D1 = Array.from({length: 2*n}, ()=>false);
    const D2 = Array.from({length: 2*n}, ()=>false);
    
    nq(0, n);
    
    function nq(x, left) {
        if (left === 0) {
            answer++;
            return;
        }
        
        for (let y=0; y<n; ++y) {
            if (R[x] || C[y] || D1[x+y] || D2[x-y]) continue;
            R[x] = true;
            C[y] = true;
            D1[x+y] = true;
            D2[x-y] = true;
            
            nq(x+1, left-1);
        
            R[x] = false;
            C[y] = false;
            D1[x+y] = false;
            D2[x-y] = false;
        }
    }
    
    return answer;
}